### ISSUE TITLE
weather widget mixing up day and night icons

### ISSUE THREAD

**Issue by lucc** [lcpz/lain#370]
I am in Germany (CET timezone) and the icon that is selected by the update function of the weather widget changes the day icon to the night icon. (I did not test it by night until now so I don't know if the night icon is also changed to the day icon but I assume it is.)

I assume the error is somewhere here: https://github.com/lcpz/lain/blob/master/widget/weather.lua#L114-L140

Sadly that code has no comments and does some crazy magic with date calculation. Git blame says that @trap000d changed these lines in #198 and #310. @trap000d can you add some comments to that code explaining what you calculate.

When I execute `curl -s 'http://api.openweathermap.org/data/2.5/weather?id=2867714&APPID=3e321f9414eaedbfab34983bda77a66e'` I get this:
```json
{
  "coord": {
    "lon": 11.58,
    "lat": 48.14
  },
  "weather": [
    {
      "id": 800,
      "main": "Clear",
      "description": "clear sky",
      "icon": "01d"
    }
  ],
  "base": "stations",
  "main": {
    "temp": 280.29,
    "pressure": 1020,
    "humidity": 56,
    "temp_min": 280.15,
    "temp_max": 281.15
  },
  "visibility": 10000,
  "wind": {
    "speed": 3.1,
    "deg": 220
  },
  "clouds": {
    "all": 0
  },
  "dt": 1512656400,
  "sys": {
    "type": 1,
    "id": 4914,
    "message": 0.0063,
    "country": "DE",
    "sunrise": 1512629438,
    "sunset": 1512660007
  },
  "id": 2867714,
  "name": "Muenchen",
  "cod": 200
}
```
Right now `date +%s` or `lua -e 'print(os.time())'` say `1512658045`. That's `Thu Dec  7 15:47:25 CET 2017` in local time or `Thu Dec  7 14:47:25 UTC 2017` and the api correctly returns the day icon. But in the widget I see the moon icon. The widgets seem to be named correctly when I view them with an image viewer.

(My /etc/localtime is a symlink to /usr/share/zoneinfo/Europe/Berlin, btw.)

**Comment by lcpz** [comment]
I am in your same timezone, and I have no problems.

Provide your `lua -v` and `awesome -v` as well.

**Comment by lucc** [comment]
My versions: `Lua 5.3.4  Copyright (C) 1994-2017 Lua.org, PUC-Rio` and 
```
awesome v4.2 (Human after all)
 • Compiled against Lua 5.3.4 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.5
 • LGI version: 0.9.2
```
and `lain-git` package from the Arch User Repository at version `1619.055e663-2`.

I took the weather script and deleted most parts to get this debug script (its just the part I linked above with print statemants):
```lua
package.path = package.path .. ';/usr/share/awesome/lib/lain/util/?.lua'
local json = require"dkjson"

local current_call = "curl -s 'http://api.openweathermap.org/data/2.5/weather?id=2867714&units=metric&lang=en&APPID=3e321f9414eaedbfab34983bda77a66e'"

local stdout = io.popen(current_call):read('*all')
print('stdout',stdout)

local weather_now = json.decode(stdout)

local sunrise = tonumber(weather_now["sys"]["sunrise"])
print('sunrise', sunrise)
local sunset  = tonumber(weather_now["sys"]["sunset"])
print('sunset', sunset)
local icon    = weather_now["weather"][1]["icon"]
print('icon',icon)
local loc_now = os.time()
print('loc_now',loc_now)
local loc_m   = os.time { year = os.date("%Y"), month = os.date("%m"), day = os.date("%d"), hour = 0 }
print('loc_m',loc_m)
local loc_t   = os.difftime(loc_now, loc_m)
print('loc_t',loc_t)
local loc_d   = os.date("*t",  loc_now)
local utc_d   = os.date("!*t", loc_now)
local utc_now = os.time(utc_d)
print('utc_now',utc_now)
local offdt   = (loc_d.isdst and 1 or 0) * 3600 + 100 * (loc_d.min  - utc_d.min) / 60
print('offdt',offdt)
local offset  = os.difftime(loc_now, utc_now) + offdt
print('offset',offset)
local offday  = (offset < 0 and -86400) or 86400
print('offday',offday)

print('if 1')
if math.abs(loc_now - utc_now - offdt + loc_t) >= 86400 then
    utc_now = utc_now + offday
end
print('utc_now',utc_now)

print('if 2')
if offday * (loc_now - utc_now - offdt) > 0 then
    sunrise = sunrise + offday
    sunset  = sunset  + offday
end
print('sunrise', sunrise)
print('sunset', sunset)

print('if 3')
if sunrise <= loc_now and loc_now <= sunset then
    icon = string.gsub(icon, "n", "d")
else
    icon = string.gsub(icon, "d", "n")
end
print('icon',icon)
```
And right now I get this output (it is 9:49 CET):
```
stdout	{"coord":{"lon":11.58,"lat":48.14},"weather":[{"id":500,"main":"Rain","description":"light rain","icon":"10d"}],"base":"stations","main":{"temp":3.62,"pressure":1010,"humidity":86,"temp_min":2,"temp_max":5},"visibility":10000,"wind":{"speed":8.2,"deg":240},"clouds":{"all":75},"dt":1512721200,"sys":{"type":1,"id":4914,"message":0.0029,"country":"DE","sunrise":1512715885,"sunset":1512746401},"id":2867714,"name":"Muenchen","cod":200}
sunrise	1512715885
sunset	1512746401
icon	10d
loc_now	1512722990
loc_m	1512687600
loc_t	35390.0
utc_now	1512719390
offdt	0.0
offset	3600.0
offday	86400
if 1
utc_now	1512719390
if 2
sunrise	1512802285
sunset	1512832801
if 3
icon	10n
```

**Comment by lucc** [comment]
I noticed that my system gives that the same number of seconds since the epoch if I ask for UTC or local time: `date ; date -u ; date +%s ; date -u +%s; lua -e 'print(os.time())'`in the shell prints 
```
Fri Dec  8 09:58:26 CET 2017
Fri Dec  8 08:58:26 UTC 2017
1512723506
1512723506
1512723506
```
and according to [wikipedia](https://en.wikipedia.org/wiki/Unix_time) this sounds correct to me.

But the weather script has a variable loc_now with a different unix timestamp than the variable utc_now (exactly 1hour difference). What is going on here? Is my system strange (how to find out? I think it is not as I just followed the standard arch linux install instructions to set the time and time zone) or does the script make some strange assumptions or are the variable names just misleading me?

**Comment by lcpz** [comment]
Thanks for the data. I'm very busy right now, but I'll check this ASAP.

**Comment by trap000d** [comment]
Hmm, looks like I've poured too much magic to the code when optimized it half a year ago. Script calculates all values properly, however result is wrong. I'll have a look at old variant and try to figure out where the logic was broken.

**Comment by mratmartinez** [comment]
Having the same problem. [This is the city where I live](http://openweathermap.org/city/3429577)

And my timezone is set to /usr/share/zoneinfo/America/Argentina/Buenos_Aires

**Comment by lcpz** [comment]
@mratmartinez https://github.com/lcpz/lain/issues/370#issuecomment-350221340

Also, every help is welcome.

**Comment by trap000d** [comment]
Sorry I was far away from home for 2 months. Here the fix, however it's not tested yet: https://github.com/trap000d/lain/commit/0b734c55e9ae37f9e12aa3ce8ffd7930c58555a2
I'm going to strip debug and create normal PR in a next few days, meanwhile you might give it a try

**Comment by lcpz** [comment]
It should be solved in b406e6f. Both @lucc and @mratmartinez timezones are correctly parsed. Please report if you (still) have (other) issues.
